alg/teca_model_segmentation.py

```diff
@@ -0,0 +1,206 @@
+                rep['variables'] = []
+
+            if self.pred_name:
+                rep['variables'].append(self.pred_name)
+
+            return rep
+        return report
+
+
+    def get_request_callback(self):
+        """
+        return a teca_algorithm::request function adding the variable name
+        that the pretrained model will process.
+        """
+        def request(port, md_in, req_in):
+            if not self.variable_name:
+                # TODO if this is part of a parallel pipeline then
+                # only rank 0 should report an error.
+                sys.stderr.write('ERROR: No variable to request specifed\n')
+                return []
+
+            req = teca_py.teca_metadata(req_in)
+
+            arrays = []
+            if req.has('arrays'):
+                arrays = req['arrays']
+            
+            arrays.append(self.variable_name)
+            req['arrays'] = arrays
+
+            return [req]
+        return request
+
+    def get_predictions_execute(self):
+        """
+        return a teca_algorithm::execute function
+        """
+        def execute(port, data_in, req):
+            """
+            expects an array of an input variable to run through
+            the torch model and get the segmentation results as an 
+            output.
+            """
+            in_mesh = teca_py.as_teca_cartesian_mesh(data_in[0])
+
+            if in_mesh is None:
+                # TODO if this is part of a parallel pipeline then
+                # only rank 0 should report an error.
+                sys.stderr.write('ERROR: empty input, or not a mesh\n')
+                return teca_py.teca_cartesian_mesh.New()
+
+            if self.model is None:
+                # TODO if this is part of a parallel pipeline then
+                # only rank 0 should report an error.
+                sys.stderr.write('ERROR: pretrained model has not been specified\n')
+                return teca_py.teca_cartesian_mesh.New()
+
+            lat = np.array(in_mesh.get_y_coordinates())
+            lon = np.array(in_mesh.get_x_coordinates())
+
+            arrays = in_mesh.get_point_arrays()
+
+            var_array = arrays[self.variable_name]
+
+            if self.transform_fn:
+                var_array = self.transform_fn(var_array, *self.transport_fn_args)
+
+            var_array = torch.from_numpy(var_array).to(self.device)
+
+            with torch.no_grad():
+                pred = F.sigmoid(self.model(var_array))
+
+            if pred is None:
+                # TODO if this is part of a parallel pipeline then
+                # only rank 0 should report an error.
+                sys.stderr.write('ERROR: Model failed to get predictions\n')
+                return teca_py.teca_cartesian_mesh.New()
+            
+            out_mesh = teca_py.teca_cartesian_mesh.New()
+            out_mesh.shallow_copy(in_mesh)
+
+            pred = teca_py.teca_variant_array.New(pred.numpy().flatten())
```

Comment:
avoid using flatten, which always makes a copy.  prefer ravel instead because it returns a view.